Allow "floating" type variables with deriveVia

Currently, `deriving-compat` errors if you try to do this:

```hs
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeOperators #-}
module Bug where

import Data.Deriving.Via
import Data.Functor.Const

newtype Age = MkAge Int
$(deriveVia [t| forall a. Show Age `Via` Const Int a |])
```
```
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )
Bug.hs:10:3-55: Splicing declarations
    deriveVia [t| forall a_a1GC. Show Age `Via` Const Int a_a1GC |]
  ======>
    instance Show Age where
      showsPrec :: Int -> Age -> ShowS
      show :: Age -> String
      showList :: [Age] -> ShowS
      showsPrec
        = (GHC.Prim.coerce
             @(Int -> Const Int a_a5gh -> ShowS) @(Int -> Age -> ShowS))
            showsPrec
      show
        = (GHC.Prim.coerce @(Const Int a_a5gh -> String) @(Age -> String))
            show
      showList
        = (GHC.Prim.coerce
             @([Const Int a_a5gh] -> ShowS) @([Age] -> ShowS))
            showList

Bug.hs:10:3: error:
    The exact Name ‘a_a5gh’ is not in scope
      Probable cause: you used a unique Template Haskell name (NameU), 
      perhaps via newName, but did not bind it
      If that's it, then -ddump-splices might be useful
   |
10 | $(deriveVia [t| forall a. Show Age `Via` Const Int a |])
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Bug.hs:10:3: error:
    The exact Name ‘a_a5gh’ is not in scope
      Probable cause: you used a unique Template Haskell name (NameU), 
      perhaps via newName, but did not bind it
      If that's it, then -ddump-splices might be useful
   |
10 | $(deriveVia [t| forall a. Show Age `Via` Const Int a |])
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Bug.hs:10:3: error:
    The exact Name ‘a_a5gh’ is not in scope
      Probable cause: you used a unique Template Haskell name (NameU), 
      perhaps via newName, but did not bind it
      If that's it, then -ddump-splices might be useful
   |
10 | $(deriveVia [t| forall a. Show Age `Via` Const Int a |])
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

However, `deriving via Const Int a instance Show Age` _does_ work on GHC HEAD after commit [055bfef](https://gitlab.haskell.org/ghc/ghc/commit/055bfefd01778c5ada8edf86c2b0a5c03d136c81). This is because GHC HEAD now generates (roughly) the following code:

```hs
    instance forall a. Show Age where
      showsPrec :: Int -> Age -> ShowS
      show :: Age -> String
      showList :: [Age] -> ShowS
      showsPrec
        = (GHC.Prim.coerce
             @(Int -> Const Int a -> ShowS) @(Int -> Age -> ShowS))
            showsPrec
      show
        = (GHC.Prim.coerce @(Const Int a -> String) @(Age -> String))
            show
      showList
        = (GHC.Prim.coerce
             @([Const Int a] -> ShowS) @([Age] -> ShowS))
            showList
```

Note the use of `forall a_a5gh` in the instance head. `deriveVia` should be able to do much the same, especially since you have to write out the `forall a` part in the `TypeQ` argument provided to `deriveVia` anyway.

Patch incoming.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,11 @@
+### next [????.??.??]
+* Permit `deriveVia` to use "floating" `via` type variables, such as the `a`
+  in:
+
+  ```hs
+  deriveVia [t| forall a. Show MyInt `Via` Const Int a |]
+  ```
+
 ### 0.5.6 [2019.05.02]
 * Support deriving `Eq`, `Ord`, and `Show` instances for data types with fields
   of type `Int8#`, `Int16#`, `Word8#`, or `Word16#` on GHC 8.8 or later.

--- a/deriving-compat.cabal
+++ b/deriving-compat.cabal
@@ -158,6 +158,7 @@ test-suite spec
                        ShowSpec
                        GH6Spec
                        GH24Spec
+                       GH27Spec
 
                        Types.EqOrd
                        Types.ReadShow

--- a/src/Data/Deriving.hs
+++ b/src/Data/Deriving.hs
@@ -79,25 +79,25 @@ The following changes have been backported:
 * In GHC 8.4, deriving 'Functor' and 'Traverable' was changed so that it uses 'coerce'
   for efficiency when the last parameter of the data type is at phantom role.
 
-* In GHC 8.4, the `EmptyDataDeriving` proposal brought forth a slew of changes related
+* In GHC 8.4, the @EmptyDataDeriving@ proposal brought forth a slew of changes related
   to how instances for empty data types (i.e., no constructors) were derived. These
   changes include:
 
-    * For derived `Eq` and `Ord` instances for empty data types, simply return
-      `True` and `EQ`, respectively, without inspecting the arguments.
+    * For derived 'Eq' and 'Ord' instances for empty data types, simply return
+      'True' and 'EQ', respectively, without inspecting the arguments.
 
-    * For derived `Read` instances for empty data types, simply return `pfail`
-      (without `parens`).
+    * For derived 'Read' instances for empty data types, simply return 'pfail'
+      (without 'parens').
 
-    * For derived `Show` instances for empty data types, inspect the argument
-      (instead of `error`ing).
+    * For derived 'Show' instances for empty data types, inspect the argument
+      (instead of 'error'ing).
 
-    * For derived `Functor` and `Traversable` instances for empty data
-      types, make `fmap` and `traverse` strict in its argument.
+    * For derived 'Functor' and 'Traversable' instances for empty data
+      types, make 'fmap' and 'traverse' strict in its argument.
 
-    * For derived `Foldable` instances, do not error on empty data types.
-      Instead, simply return the folded state (for `foldr`) or `mempty` (for
-      `foldMap`), without inspecting the arguments.
+    * For derived 'Foldable' instances, do not error on empty data types.
+      Instead, simply return the folded state (for 'foldr') or 'mempty' (for
+      'foldMap'), without inspecting the arguments.
 
 * In GHC 8.6, the @DerivingVia@ language extension was introduced.
   @deriving-compat@ provides an interface which attempts to mimic this
@@ -107,8 +107,11 @@ The following changes have been backported:
   Since the generated code requires the use of @TypeApplications@, this can
   only be backported back to GHC 8.2.
 
-* In GHC 8.6, deriving `Read` was changed so as to factor out certain commonly
+* In GHC 8.6, deriving 'Read' was changed so as to factor out certain commonly
   used subexpressions, which significantly improve compliation times.
+
+* In GHC 8.10, @DerivingVia@ permits \"floating\" type variables in @via@ types,
+  such as the @a@ in @'deriveVia' [t| forall a. Show MyInt ``Via`` Const Int a |]@.
 -}
 
 {- $derive

--- a/src/Data/Deriving/Via/Internal.hs
+++ b/src/Data/Deriving/Via/Internal.hs
@@ -45,11 +45,12 @@ $('deriveGND' [t| forall a. 'Eq' a => 'Eq' (Foo a) |])
 deriveGND :: Q Type -> Q [Dec]
 deriveGND qty = do
   ty <- qty
-  let (instanceCxt, instanceTy) = decomposeType ty
+  let (instanceTvbs, instanceCxt, instanceTy) = decomposeType ty
   instanceTy' <- (resolveTypeSynonyms <=< resolveInfixT) instanceTy
   decs <- deriveViaDecs instanceTy' Nothing
-  (:[]) `fmap` instanceD (return instanceCxt)
-                         (return instanceTy)
+  let instanceHeader = ForallT instanceTvbs instanceCxt instanceTy
+  (:[]) `fmap` instanceD (return [])
+                         (return instanceHeader)
                          (map return decs)
 
 {- | Generates an instance for a type class by emulating the behavior of the
@@ -69,7 +70,7 @@ correctly across all the types being used (e.g., to make sure that the same
 deriveVia :: Q Type -> Q [Dec]
 deriveVia qty = do
   ty <- qty
-  let (instanceCxt, viaApp) = decomposeType ty
+  let (instanceTvbs, instanceCxt, viaApp) = decomposeType ty
   viaApp' <- (resolveTypeSynonyms <=< resolveInfixT) viaApp
   (instanceTy, viaTy)
     <- case unapplyTy viaApp' of
@@ -82,8 +83,9 @@ deriveVia qty = do
                 , "\t[t| forall a. C (T a) `Via` V a |]"
                 ]
   decs <- deriveViaDecs instanceTy (Just viaTy)
-  (:[]) `fmap` instanceD (return instanceCxt)
-                         (return instanceTy)
+  let instanceHeader = ForallT instanceTvbs instanceCxt instanceTy
+  (:[]) `fmap` instanceD (return [])
+                         (return instanceHeader)
                          (map return decs)
 
 deriveViaDecs :: Type       -- ^ The instance head (e.g., @Eq (Foo a)@)
@@ -197,9 +199,9 @@ stripOuterForallT (ForallT _ _ ty) = ty
 #endif
 stripOuterForallT ty               = ty
 
-decomposeType :: Type -> (Cxt, Type)
-decomposeType (ForallT _ ctxt ty) = (ctxt, ty)
-decomposeType ty                  = ([],   ty)
+decomposeType :: Type -> ([TyVarBndr], Cxt, Type)
+decomposeType (ForallT tvbs ctxt ty) = (tvbs, ctxt, ty)
+decomposeType ty                     = ([],   [],   ty)
 
 newtypeRepType :: DatatypeVariant -> [ConstructorInfo] -> Maybe Type
 newtypeRepType dv cons = do

--- a/tests/GH27Spec.hs
+++ b/tests/GH27Spec.hs
@@ -1,0 +1,45 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeOperators #-}
+
+#if MIN_VERSION_template_haskell(2,12,0)
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-unused-foralls #-} -- Due to GHC #13512. Sigh.
+#endif
+
+{-|
+Module:      GH27Spec
+Copyright:   (C) 2019 Ryan Scott
+License:     BSD-style (see the file LICENSE)
+Maintainer:  Ryan Scott
+Portability: Template Haskell
+
+A regression test for
+https://github.com/haskell-compat/deriving-compat/issues/27.
+We isolate this test case from the rest of "DerivingViaSpec" because
+we have to disable @-Wunused-foralls@ due to
+https://gitlab.haskell.org/ghc/ghc/issues/13512, and we would prefer not to
+contaminate the rest of "DerivingViaSpec" with this hack.
+-}
+module GH27Spec where
+
+import Prelude ()
+import Prelude.Compat
+
+import Test.Hspec
+
+#if MIN_VERSION_template_haskell(2,12,0)
+import Data.Deriving.Via
+import Data.Functor.Const
+
+newtype Age = MkAge Int
+$(deriveVia [t| forall a. Show Age `Via` Const Int a |])
+#endif
+
+main :: IO ()
+main = hspec spec
+
+spec :: Spec
+spec = pure ()

